if a call instruction was never called, remove it

o- opened the pull request:
this prevents a dead call instruction (with no call target feedback)
from poisoning the analysis, since a generic call is emited.

o- committed:
if a call instruction was never called, remove it

this prevents a dead call instruction (with no call target feedback)
from poisoning the analysis, since a generic call is emited.

charig commented on rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp:
```diff
@@ -1043,6 +1058,19 @@ Value* Rir2Pir::tryTranslate(rir::Code* srcCode, Builder& insert) const {
+            if (!inPromise() && !insert.getCurrentBB()->isEmpty()) {
+                auto last = insert.getCurrentBB()->last();
+
+                if (Deopt::Cast(last)) {
```
Isn't the break above on line 335 enough?

o- replied:
no, break just leaves the compileBC  function. we also must abort all later instructions to avoid compiling dead code.

charig commented on rir/src/compiler/translations/rir_2_pir/rir_2_pir.cpp:
```diff
@@ -320,6 +320,21 @@ bool Rir2Pir::compileBC(const BC& bc, Opcode* pos, Opcode* nextPos,
+            // unconditional deopt
+            if (!inPromise() && srcCode->funInvocationCount > 0 &&
+                feedback.taken == 0) {
+                // Hack to record the deoptimized call, to ensure we compile
```
Are we using this hack anywhere?

o- replied:
yes, without this for example binary tree which has something like:
```
tree = function(n, ...)
  if (n == 1)
    c(...)
  else
    tree(n-1, ...)
```
will optimize before the then-branch is executed. it deoptimizes on the then branch, but before it every ends up in another leaf branch, it will have optimized the function again. thus we will never record the `c` call and will always deoptimize there. the hack ensures that the `record_call` instruction is executed on deoptimization, thus on the next optimization the code will be correct.

o- commented:
@charig you can also wait with reviewing until travis is green to avoid unnecessary work. i am still not sure if this version will now pass...

o- committed:
replace broken phi input fixup after constant folding

since now all phi's are canonical, it is trivial to remove dead phi
inputs.

o- committed:
prevent deopt loops

since we are deleting deoptimized versions, we need to avoid deopt
loops.

o- committed:
only speculate on binops with typefeedback

so far we assumed non objects on types which have no typefeedback. This
leads to deopt-loops in the tests, so I changed it.

but to avoid regressions, we need to become better at using
typefeedback. As a first step I propagate the typefeedback when we
replace instructions. Ie. if we replace a for b, I copy the typefeedback
of a to b. This might not always be correct, but seems like a decent
hack for the kind of transformations we are doing mostly.

o- committed:
fix a bug where the delaying of environment into deopt branch might kill
the missing flag on a binding
